doc/developer/platform/reconciliation.md

```diff
@@ -0,0 +1,64 @@
+# Reconciliation of the ADAPTER/COMPUTE communication
+
+The current proposal for Materialize's platform includes the separation of the ADAPTER and COMPUTE components into distinct processes.
+Each represents its own failure domain and hence we need to design the communication protocols around failures and active replication (AR).
+In general, we need to define how commands and responses can be reconciled.
+
+## Command reconciliation
+
+Commands represent instructions the COMPUTE controller sends to COMPUTE instances.
+* The COMPUTE controller may restart independently of a COMPUTE instance.
+* Reconciling commands is the mechanism to match a COMPUTE instance's state to what the COMPUTE controller expects after startup.
+* On startup, the COMPUTE controller provides the current configuration to the COMPUTE instances.
+* COMPUTE instances can match the configuration to their own state and only apply changes.
+
+For this purpose, we interpret the commands COMPUTE controller provides to COMPUTE as an append-only log.
+Upon COMPUTE controller restart, the new configuration is/should be a suffix of the log of all previously received commands.
+
+The log can be compacted by collapsing commands that semantically supersede previous commands.
+
+Currently, the `Command` protocol does not express enough information to provide full command reconciliation.
```

Comment:
This might be true of `Command`, but I think is not the case for `ComputeCommand`, whose `CreateInstance` command indicates the truncation and restarting of the command log. We can try and work out `StorageCommand` reconciliation here too, but it might be out of scope / worth differentiating from `ComputeCommand` reconciliation. In particular, it isn't clear whether we end up with a foreign storage instance outside the controller that receives a sequence of `StorageCommand`s, or whether the storage controller more directly pilots its resources (e.g. containers that read and write on the data plane).